packages/server/modules/workspaces/repositories/workspaces.ts

```diff
@@ -35,26 +38,92 @@ export const upsertWorkspaceFactory =
       .merge(['description', 'logoUrl', 'name', 'updatedAt'])
   }
 
-export const getWorkspaceRoleFactory =
-  ({ db }: { db: Knex }): GetWorkspaceRole =>
+export const getWorkspaceRolesFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRoles =>
+  async ({ workspaceId }) => {
+    return await tables.workspacesAcl(db).select('*').where({ workspaceId })
+  }
+
+export const getWorkspaceRoleForUserFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRoleForUser =>
   async ({ userId, workspaceId }) => {
-    const acl = await tables
-      .workspacesAcl(db)
-      .select('*')
-      .where({ userId, workspaceId })
-      .first()
+    return (
+      (await tables
+        .workspacesAcl(db)
+        .select('*')
+        .where({ userId, workspaceId })
+        .first()) ?? null
+    )
+  }
 
-    return acl || null
+export const getWorkspaceRolesForUserFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRolesForUser =>
+  async ({ userId }, options) => {
+    const workspaceIdFilter = options?.workspaceIdFilter ?? []
+
+    const query = tables.workspacesAcl(db).select('*').where({ userId })
+
+    if (workspaceIdFilter.length > 0) {
+      query.whereIn('workspaceId', workspaceIdFilter)
+    }
+
+    return await query
+  }
+
+const isUserLastWorkspaceAdmin = (
+  workspaceRoles: WorkspaceAcl[],
+  userId: string
+): boolean => {
+  const workspaceAdmins = workspaceRoles.filter(
+    ({ role }) => role === 'workspace:admin'
+  )
+  const isUserAdmin = workspaceAdmins.some((role) => role.userId === userId)
+
+  return isUserAdmin && workspaceAdmins.length === 1
+}
+
+export const deleteWorkspaceRoleFactory =
+  ({ db }: { db: Knex }): DeleteWorkspaceRole =>
+  async ({ userId, workspaceId }) => {
+    // Protect against removing last admin in workspace
+    const workspaceRoles = await getWorkspaceRolesFactory({ db })({ workspaceId })
+
+    if (isUserLastWorkspaceAdmin(workspaceRoles, userId)) {
+      throw new Error('Cannot remove last admin in workspace.')
+    }
+
+    // Bail early if user has no role
+    const currentRole = workspaceRoles.find((role) => role.userId === userId)
+
+    if (!currentRole) {
+      return null
+    }
+
+    // Perform delete
+    await tables.workspacesAcl(db).where({ workspaceId, userId }).delete()
+
+    return currentRole ?? null
   }
 
 export const upsertWorkspaceRoleFactory =
   ({ db }: { db: Knex }): UpsertWorkspaceRole =>
   async ({ userId, workspaceId, role }) => {
+    // Verify requested role is valid workspace role
     const validRoles = Object.values(Roles.Workspace)
     if (!validRoles.includes(role)) {
       throw new Error(`Unexpected workspace role provided: ${role}`)
     }
 
+    // Protect against removing last admin in workspace
+    const workspaceRoles = await getWorkspaceRolesFactory({ db })({ workspaceId })
```

Comment:
Now this the name of the function is lying to me. This func is more like `ensuringAnAdminRemainsAndUpsertsTheRoleFactory`

I think this is a lot cleaner as service level guarantee, its an application logic decision, that we do not allow workspaces without admin roles. The repo itself should not implement these guards. If we want that to be a strict dataschema guard, we should use database level guards instead..